### PR TITLE
add some e2e for node authz

### LINKED ISSUE #47174: e2e test the Node Authorizer
**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):

TESTING REQUEST

Spawned from: https://github.com/kubernetes/kubernetes/pull/46879
It would be nice with some e2e tests for the Node Authorizer so consumers of it (kubeadm, GCE, GKE, etc.) could easily make sure it's enabled and working, just by editing `ginkgo.focus` when running e2e's.

I wrote a quick bash script of what it could do:

```bash
if [[ $(kubectl version 2>&1 | grep "Client" | grep 'Minor:"7') == "" ]]; then
	echo "kubectl v1.7.0-beta.0 or higher is required"
	exit 1
fi

export KUBECONFIG=/etc/kubernetes/admin.conf
IMPERSONATE="--as system:node:thegopher --as-group system:nodes"

printf "TESTING: Getting a non-existent secret should exit with the Forbidden error, not a NotFound error] "
([[ $(kubectl ${IMPERSONATE} get secret foo 2>&1 | grep "Forbidden") != "" ]] && printf "PASSED!\n") || printf "FAILED!\n"

printf "TESTING: Getting an existent secret should exit with the Forbidden error] "
([[ $(kubectl ${IMPERSONATE} get secret $(kubectl get secret | grep token | awk '{print $1}') 2>&1 | grep "Forbidden") != "" ]] && printf "PASSED!\n") || printf "FAILED!\n"

printf "TESTING: Getting a secret for a workload the node has access to should succeed] "
kubectl ${IMPERSONATE} -n kube-system get secrets $(kubectl -n kube-system get secret | grep kube-proxy | awk '{print $1}') 1>/dev/null 2>/dev/null
([[ $? == 0 ]] && printf "PASSED!\n") || printf "FAILED!\n"

printf "TESTING: A node shouldn't be able to create an other node] "
echo '{"apiVersion":"v1","kind":"Node","metadata":{"name":"foo"}}' | kubectl delete -f - 1>/dev/null 2>/dev/null
echo '{"apiVersion":"v1","kind":"Node","metadata":{"name":"foo"}}' | kubectl ${IMPERSONATE} --validate=false create -f - 1>/dev/null 2>/dev/null
([[ $? == 1 ]] && printf "PASSED!\n") || printf "FAILED!\n"
```

The bash script above is in no way comprehensive enough, but on the other hand it took  ~5mins to write :)

For now (1.7), we might use something like this bash script in our kubeadm e2e CI job to make sure the Node Authorizer/Admission plugin is enabled correctly.

cc @kubernetes/sig-cluster-lifecycle-feature-requests @kubernetes/sig-auth-feature-requests

### PR DESCRIPTION
**Pull request by wanghaoran1988**
**What this PR does / why we need it**:
fix #47174
**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #47174

**Special notes for your reviewer**:

**Release note**:
```
None
```

### CODE REVIEW

**Review by luxas — CHANGES_REQUESTED** [review]
Thanks @wanghaoran1988 for working on this!

When I wrote up the issue for this, I imagined that this would actually be written using the normal Go client, not exec out to kubectl. The kubectl commands I had there were for PoC and reference of something that could be implemented.

What does @kubernetes/sig-auth-pr-reviews think?
My opinion is that this should use the Go client as pretty much everything else.

**Comment by luxas** [comment]
Please fix up bazel/verify/unit tests as well

**Comment by ericchiang** [comment]
@wanghaoran1988 client-go has impersonation config that can be used to act as arbitrary users in the system. Might be a good place to start https://godoc.org/k8s.io/client-go/rest#Config

You can do actual inspection of the error codes that way and avoid patterns like inspecting the stderr.

```
Expect(err).ShouldNot(ContainSubstring("not found"))
```

**Comment by wanghaoran1988** [comment]
@ericchiang Thanks for your suggestions,  @luxas  Updated using the rest client, PTAL

File: test/e2e/auth/node_authz.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+	var defaultSaSecret string
+	var nodeName string
+	BeforeEach(func() {
+		ns = f.Namespace.Name
+
+		nodeList, err := f.ClientSet.CoreV1().Nodes().List(metav1.ListOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		Expect(len(nodeList.Items)).NotTo(Equal(0))
+		nodeName = nodeList.Items[0].Name
+		asUser = "system:node:" + nodeName
```
Comment (luxas) [inline test/e2e/auth/node_authz.go]: isn't there any helper function we can already use here for this task?

File: test/e2e/auth/node_authz.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+	. "github.com/onsi/gomega"
+)
+
+var _ = SIGDescribe("[Feature:NodeAuthz]", func() {
+
+	f := framework.NewDefaultFramework("node-authz")
+	var c clientset.Interface
+	var ns string
+	var asUser string
+	var asGroup string = "system:nodes"
```
Comment (luxas) [inline test/e2e/auth/node_authz.go]: I'm pretty sure this exists as a constant somewhere

File: test/e2e/auth/node_authz.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/kubernetes/test/e2e/framework"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = SIGDescribe("[Feature:NodeAuthz]", func() {
```
Comment (luxas) [inline test/e2e/auth/node_authz.go]: nit: spell out the whole `NodeAuthorizer`

File: test/e2e/auth/node_authz.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+		nodeList, err := f.ClientSet.CoreV1().Nodes().List(metav1.ListOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		Expect(len(nodeList.Items)).NotTo(Equal(0))
+		nodeName = nodeList.Items[0].Name
+		asUser = "system:node:" + nodeName
+		sa, err := f.ClientSet.CoreV1().ServiceAccounts(ns).Get("default", metav1.GetOptions{})
+		Expect(len(sa.Secrets)).NotTo(Equal(0))
+		Expect(err).NotTo(HaveOccurred())
+		defaultSaSecret = sa.Secrets[0].Name
+		By("Creating a kubernetes client that impersonate a node")
```
Comment (luxas) [inline test/e2e/auth/node_authz.go]: nit: impersonates a node

File: test/e2e/auth/node_authz.go [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+	"k8s.io/kubernetes/test/e2e/framework"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = SIGDescribe("[Feature:NodeAuthz]", func() {
+
+	f := framework.NewDefaultFramework("node-authz")
+	var c clientset.Interface
```
Comment (luxas) [inline test/e2e/auth/node_authz.go]: should this be more describing, like `nodeImpersonatedClient`? That's probably too long...
Maybe just add a comment above `c` about what it does...

File: test/e2e/auth/node_authz.go [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+			},
+		}
+		f.ClientSet.CoreV1().Pods(ns).Create(pod)
+		err := f.WaitForPodRunning("pause")
+		Expect(err).NotTo(HaveOccurred())
+		_, err = c.CoreV1().Secrets(ns).Get(defaultSaSecret, metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	It("A node shouldn't be able to create/delete an other node", func() {
```
Comment (luxas) [inline test/e2e/auth/node_authz.go]: Can we split this in two separate cases please?

File: test/e2e/auth/node_authz.go [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+				Containers: []v1.Container{
+					{
+						Name:  "pause",
+						Image: framework.GetPauseImageName(f.ClientSet),
+					},
+				},
+				NodeName: nodeName,
+			},
+		}
+		f.ClientSet.CoreV1().Pods(ns).Create(pod)
```
Comment (luxas) [inline test/e2e/auth/node_authz.go]: maybe you should return err here and add one more `Expect(err).NotTo(HaveOccurred())`?

File: test/e2e/auth/framework.go [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package auth
+
+import "github.com/onsi/ginkgo"
+
+func SIGDescribe(text string, body func()) bool {
```
Comment (luxas) [inline test/e2e/auth/framework.go]: Add `// SIGDescribe prepends the name of the current SIG (sig-auth) that is owning the e2e tests in this folder` or something like that

File: hack/.golint_failures [hunk 9]
```diff
@@ -801,6 +801,7 @@ staging/src/k8s.io/sample-apiserver/pkg/registry/wardle/flunder
 test/e2e
 test/e2e/apimachinery
 test/e2e/apps
+test/e2e/auth
```
Comment (luxas) [inline hack/.golint_failures]: Can you please make golint pass for the new code here or is there anything blocking that?

**Review by luxas — APPROVED** [review]
Thanks @wanghaoran1988!
Overall looks great, increases our coverage drastically compared to what we have now (exactly 0 e2e tests on this) :smile:

I had some nits, please fixup and I'll re-review

File: hack/.golint_failures (hunk 9 above)
Comment (wanghaoran1988) [inline hack/.golint_failures]: I'd like to add here like other pkg, one problem is that golint do not allow dot impots like:
```
. "github.com/onsi/ginkgo"
. "github.com/onsi/gomega"
```

File: test/e2e/auth/node_authz.go (hunk 6 above)
Comment (wanghaoran1988) [inline test/e2e/auth/node_authz.go]: sure

File: test/e2e/auth/node_authz.go (hunk 7 above)
Comment (wanghaoran1988) [inline test/e2e/auth/node_authz.go]: good catch

File: hack/.golint_failures (hunk 9 above)
Comment (luxas) [inline hack/.golint_failures]: Ah, ok, then I see why all other e2e's are here. Didn't know golint was so picky.

**Review by luxas — APPROVED** [review]
/lgtm
Thanks!

**Comment by luxas** [comment]
cc @kubernetes/sig-auth-pr-reviews 
PTAL @liggitt

File: test/e2e/auth/node_authz.go [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+						Name:  "pause",
+						Image: framework.GetPauseImageName(f.ClientSet),
+					},
+				},
+				NodeName: nodeName,
+			},
+		}
+		_, err := f.ClientSet.CoreV1().Pods(ns).Create(pod)
+		Expect(err).NotTo(HaveOccurred())
+		err = f.WaitForPodRunning("pause")
```
Comment (liggitt) [inline test/e2e/auth/node_authz.go]: the pod doesn't have to be running to verify the node has access

File: test/e2e/auth/node_authz.go [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+		Expect(err).NotTo(HaveOccurred())
+		_, err = c.CoreV1().Secrets(ns).Get(defaultSaSecret, metav1.GetOptions{})
```
Comment (liggitt) [inline test/e2e/auth/node_authz.go]: should put this in a retry loop... access depends on the authorizer graph observing the pod

File: test/e2e/auth/node_authz.go [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+		Expect(apierrors.IsForbidden(err)).Should(Equal(true))
+	})
+
+	It("Getting an existent secret should exit with the Forbidden error", func() {
+		_, err := c.CoreV1().Secrets(ns).Get(defaultSaSecret, metav1.GetOptions{})
+		Expect(apierrors.IsForbidden(err)).Should(Equal(true))
+	})
+
+	It("Getting a secret for a workload the node has access to should succeed", func() {
+		pod := &v1.Pod{
```
Comment (liggitt) [inline test/e2e/auth/node_authz.go]: I'm not sure how ginkgo parallelizes tests, but this should probably be more of a standalone test, which means creating a secret specifically for this test, ensuring the node has no access, then creating the pod and ensuring the node has access.

File: test/e2e/auth/node_authz.go [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+const (
+	NodesGroup     = "system:nodes"
+	NodeNamePrefix = "system:node:"
+)
+
+var _ = SIGDescribe("[Feature:NodeAuthorizer]", func() {
```
Comment (liggitt) [inline test/e2e/auth/node_authz.go]: do any tests actually enable this feature? if not, this hasn't really increased our coverage.

**Comment by liggitt** [comment]
https://github.com/kubernetes/kubernetes/pull/48966 moves auth-related e2e's into a subpackage... I would wait until that merges, then rebase on master

File: test/e2e/auth/node_authz.go (hunk 13 above)
Comment (luxas) [inline test/e2e/auth/node_authz.go]: @liggitt the kubeadm job will right after this is merged

**Comment by ericchiang** [comment]
https://github.com/kubernetes/kubernetes/pull/48966 has now merged so I think this PR is unblocked once the merge conflicts are resolved.

**Comment by wanghaoran1988** [comment]
@luxas @liggitt PR rebased and addressed the comments

File: test/e2e/auth/node_authz.go [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+						Image: framework.GetPauseImageName(f.ClientSet),
+					},
+				},
+				NodeName: nodeName,
+				Volumes: []v1.Volume{
+					{
+						Name: "node-auth-secret",
+						VolumeSource: v1.VolumeSource{
+							Secret: &v1.SecretVolumeSource{
+								SecretName: "node-auth-secret",
```
Comment (luxas) [inline test/e2e/auth/node_authz.go]: this should be `secret.ObjectMeta.Name`

File: test/e2e/auth/node_authz.go [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+				},
+			},
+		}
+
+		_, err = f.ClientSet.CoreV1().Pods(ns).Create(pod)
+		Expect(err).NotTo(HaveOccurred())
+
+		By("The node should able to access the secret")
+		err = wait.Poll(framework.Poll, 1*time.Minute, func() (bool, error) {
+			_, err = c.CoreV1().Secrets(ns).Get(defaultSaSecret, metav1.GetOptions{})
```
Comment (luxas) [inline test/e2e/auth/node_authz.go]: this should be `secret.ObjectMeta.Name`

**Review by luxas — COMMENTED** [review]
Two comments that should be fixed before merge yet

File: test/e2e/auth/node_authz.go [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+			if err != nil {
+				framework.Failf("Failed to get secret %v, err: %v", defaultSaSecret, err)
+				return false, err
```
Comment (ericchiang) [inline test/e2e/auth/node_authz.go]: Returning here will cause wait.Poll to exit with a non-nil error code. Same with Failf. Isn't the intent to retry after a first initial failure?

This should probably just log the error and `return false, nil`, right?

Comment (wanghaoran1988) [inline test/e2e/auth/node_authz.go]: Good catch, thanks.

**Review by luxas — APPROVED** [review]
/lgtm

File: test/e2e/auth/node_authz.go [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+		}
+
+		_, err = f.ClientSet.CoreV1().Pods(ns).Create(pod)
+		Expect(err).NotTo(HaveOccurred())
+
+		By("The node should able to access the secret")
+		err = wait.Poll(framework.Poll, 1*time.Minute, func() (bool, error) {
+			_, err = c.CoreV1().Secrets(ns).Get(secret.Name, metav1.GetOptions{})
+			if err != nil {
+				framework.Failf("Failed to get secret %v, err: %v", secret.Name, err)
```
Comment (ericchiang) [inline test/e2e/auth/node_authz.go]: This is still Failf

Comment (luxas) [inline test/e2e/auth/node_authz.go]: Yeah, that'd be a bug. Sharp eye. @wanghaoran1988 please fix

**Comment by ericchiang** [comment]
Approving. Please feel free to lgtm once that bug is fixed.

/approve

**Comment by wanghaoran1988** [comment]
fixed

**Comment by luxas** [comment]
/lgtm

Let's get this enabled for kubeadm and we'll see how it goes :)

**Pull request merged**